### ISSUE TITLE
Feature request: support loading fonts with raw data

### ISSUE THREAD

**Issue by homerjam** [thx/resvg-js#101]
Hello, thanks for this awesome library - I came across it via svg2img.

Would it be possible to support loading fonts via raw data? It's seems this might be possible using the font library dependency https://github.com/RazrFalcon/fontdb#features

The reason is that on Vercel I can't seem to use fonts because file system access is protected/complicated. I've also tried embedding the font in the svg with font-face/data-uri but this doesn't seem to be supported either.

Thanks

**Comment by zimond** [comment]
Yes, this feature is possible and scheduled. We'll find some time to implement it soon.

**Comment by snelsi** [comment]
@zimond 
Hey 👋
Any updates on this?
I'm currently struggling to load fonts to my svg as well

**Comment by yisibl** [comment]
@snelsi This work is very complex and @zimond is being worked on.

**Comment by snelsi** [comment]
@yisibl @zimond 
Do you have any rough estimates of when this feature will be released?
Thank you!

**Comment by homerjam** [comment]
> @yisibl @zimond Do you have any rough estimates of when this feature will be released? Thank you!

@snelsi just so you know for the future, to illicit a positive response a better way to phase this question would be "is there anything I can do to help with this feature such as providing ideas, coding, testing or monetary support?"

**Comment by zimond** [comment]
Never mind. It indeed takes longer time than estimated. There's an open PR #118 addressing this feature. It works now it you manually built the artifacts. So if this is what you really want, you could try the pr and any feedback is welcomed as it dramatically changed the current API design.

**Comment by filipedeschamps** [comment]
Hi everyone! I'm having the exact same problem. Everything works on Vercel, [except the fonts](https://tabnews-git-dynamic-thumbnails-tabnews.vercel.app/api/v1/contents/filipedeschamps/testando-deploy-com-nova-regra-de-slug-sobre-under-score-e-tambem-a-e-b-a-e-b/thumbnail).

Has anyone found a workaround?

**Comment by filipedeschamps** [comment]
I found a way to make it work on Vercel 🤝  the above link is working and should show you this:

![image](https://user-images.githubusercontent.com/4248081/180319571-d3a49088-d2e5-4b3b-98c5-5f3943c2bf58.png)

Inside your `next.config.js` you should add:

```js
  experimental: {
    nftTracing: true,
  },
```

For example, my current `next.config.js`

```js
module.exports = {
  pageExtensions: ['public.js'],
  eslint: {
    ignoreDuringBuilds: true,
  },
  experimental: {
    nftTracing: true,
  },
  compiler: {
    styledComponents: true,
  },
  i18n: {
    locales: ['pt-br'],
    defaultLocale: 'pt-br',
  },
};
```

And then, you can resolve the file path like this:

```js
const { join, resolve } = require('path');

// ...

  const result = await renderAsync(svg, {
    fitTo: {
      mode: 'width',
      value: 1280,
    },
    font: {
      fontFiles: [
        join(resolve('.'), 'fonts', 'Roboto-Regular.ttf'),
        join(resolve('.'), 'fonts', 'Roboto-Bold.ttf'),
        join(resolve('.'), 'fonts', 'NotoEmoji-Bold.ttf'),
      ],
      loadSystemFonts: false,
      defaultFontFamily: 'Roboto',
    },
  });
```

## But, there's still one problem

Despite that it works 100% on Vercel, I can't make it work on **Github Actions**.

@homerjam is there any way to make the module more `verbose`? Thank you.

**Comment by zernonia** [comment]
@filipedeschamps do you host the fonts locally? and how did you apply font to each text? `font-family="Roboto"`?

**Comment by filipedeschamps** [comment]
> do you host the fonts locally?

Yes, in the path `/fonts/Roboto-Regular.ttf` for example. That's why I can resolve them like this:

```js
join(resolve('.'), 'fonts', 'Roboto-Regular.ttf'),
```

> and how did you apply font to each text? `font-family="Roboto"`?

No, it's being applied globally:

```js
defaultFontFamily: 'Roboto',
```

**Comment by zimond** [comment]
I'm sorry this is currently halted. Maybe #217  could fit your needs? If so we will try to merge that PR this week.

**Comment by Pondorasti** [comment]
I think that would do it @zimond! Looking forward to the #217 changes 🤗

**Comment by arkaydeus** [comment]
> Any updates on this? I've tried @filipedeschamps solution, but without any success.
> 
> My font is located at `public/_static/font/Inter-VariableFont_slnt,wght.ttf`. I've enabled `outputFileTracing: true` in `next.config.js`, and then tried creating the path using both `resolve('.')` and `process.cwd()`, but neither solutions worked.
> 
> ```
> font: {
>   fontFiles: [resolve('.'), "_static/font", "Inter-VariableFont_slnt,wght.ttf")],
>   // fontFiles: [join(process.cwd(), "_static/font", "Inter-VariableFont_slnt,wght.ttf")],
>   defaultFontFamily: "Inter",
>   loadSystemFonts: false,
> }
> ```
> 
> I've also tried forgetting about custom fonts and just using system fonts but that didn't work either. On local machine everything is fine, but on vercel deployments text is not rendered.
> 
> ```
> font: {
>   loadSystemFonts: true,
> }
> ```
> 
> During local development I just used to have an absolute path, which works perfectly fine, but obviously breaks during deployments...
> 
> ```
> font: {
>   fontFiles: ["./public/_static/font/Inter-VariableFont_slnt,wght.ttf"],
>   defaultFontFamily: "Inter",
>   loadSystemFonts: false,
> }
> ```
> 
> I've also inspected build output in Vercel and I can see my font file as part of it, but none of the path resolutions from above would work during deployments.
> 
> ![CleanShot 2023-04-10 at 12 10 29](https://user-images.githubusercontent.com/32957606/230978483-4988b64b-561d-4480-b186-b4f80d1114fa.png)

Did you manage to find a solution in the end?

**Comment by arkaydeus** [comment]
So I've confirmed that the Vercel build does have my fonts in the fonts folder:

<img width="299" alt="image" src="https://user-images.githubusercontent.com/39967247/235307275-c8346cfc-e7a1-406a-bf72-ec614231d530.png">

This is the font section of my opts:

```js  
font: {
    fontFiles: [
      join(resolve('.'), 'fonts', 'OpenSans-Medium.ttf'),
      join(resolve('.'), 'fonts', 'RobotoMono-Regular')
    ], // Load custom fonts.
    // fontFiles: ['fonts/RobotoMono-Regular.ttf'],
    loadSystemFonts: false, // It will be faster to disable loading system fonts.
    defaultFontFamily: 'Roboto Mono' // Set default font family.
  }
```

next.config.js is as follows:

```js
const nextConfig = {
  reactStrictMode: true,
  experimental: {
    nftTracing: true
  },
  outputFileTracing: true
}
```

Everything works fine locally but not on Vercel. Any ideas?

**Comment by Pondorasti** [comment]
thanks for sharing your findings @arkaydeus, I've got it working too 🤗! Looks like the missing piece was to place my fonts at root instead of the `/static` directory.

> Put the fonts in the root of your repo in a folder called fonts. Strangely this does not show up in the build output.

No kidding, it breaks the Vercel dashboard output page for me. 🤣


---

Btw, `nftTracing` is not really needed. It's not even part of the config spec if you look at the typings. I believe `nftTracing` got renamed to `outputFileTracing` when it made it out of experimental.


```diff
const nextConfig = {
   reactStrictMode: true,
-  experimental: {
-    nftTracing: true
-  },
   outputFileTracing: true
}
```
